File: correct_velocity.m
function Shot = correct_velocity(Shot)

disp(Shot)

% we try to correct noise and velocity jumps
% sweep over several points and look for velocity changes
% if velocity change drop -X % is detected activate
% if in next 5 timesteps a velocity
%     increase +XX% & decrease to old value is with tolerance
% is detected, then reposition position values based on new velocities
tsteps = length(Shot.t);
dx = diff(Shot.x);
dy = diff(Shot.y);
dt = diff(Shot.t);

ds = sqrt((Shot.x(2:end)-Shot.x(1:end-1)).^2 + (Shot.y(2:end) - Shot.y(1:end-1)).^2)/1000;
vel = ds./dt;

ind = find(dt > mean(dt));

figure
plot( Shot.t, Shot.x,'-')
hold on
grid on
plot( Shot.t, Shot.y,'-')

plot( Shot.t(ind), Shot.y(ind),'o')

check = 1;
while check
    
    %% Identify single extremes
    dev = vel_mvmean - vel;
    ind = find(dev < -0.5,1,'first');
    
    if ~isempty(ind)
        subplot(3,1,1)
        plot( Shot.t(ind), dx(ind), 'o')
        plot( Shot.t(ind), dy(ind), 'o')
        
        subplot(3,1,2)
        plot( Shot.t(ind), vel(ind), 'o')
        
        subplot(3,1,3)
        plot( Shot.t(ind), Shot.x(ind), 'o')
        plot( Shot.t(ind), Shot.y(ind), 'o')
        
        if abs((dx(ind+1)-dx(ind-1))/dx(ind-1)) < 0.1 | abs((dy(ind+1)-dy(ind-1))/dy(ind-1)) < 0.1
            dx1 = (dx(ind+1)+dx(ind-1))/2;
            dy1 = (dy(ind+1)+dy(ind-1))/2;
            Shot.x(ind) = Shot.x(ind-1) + dx1;
            Shot.y(ind) = Shot.y(ind-1) + dy1;
            
            dx = diff(Shot.x);
            dy = diff(Shot.y);
            
            ds = sqrt((Shot.x(2:end)-Shot.x(1:end-1)).^2 + (Shot.y(2:end) - Shot.y(1:end-1)).^2)/1000;
            dt = diff(Shot.t);
            vel = ds./dt;
            
            subplot(3,1,1)
            plot( Shot.t(ind), dx(ind), 'x')
            plot( Shot.t(ind), dy(ind), 'x')
            
            subplot(3,1,2)
            plot( Shot.t(ind), vel(ind), 'x')
            
            subplot(3,1,3)
            plot( Shot.t(ind), Shot.x(ind), 'x')
            plot( Shot.t(ind), Shot.y(ind), 'x')
            
            
            figure
            plot( Shot.t, Shot.y, '-x')
        end
    end
    
end

File: correct_velocity.py
import numpy as np
import matplotlib.pyplot as plt

def correct_velocity(shot):
    print(shot)

    # Calculate differences and velocities
    dx = np.diff(shot['x'])
    dy = np.diff(shot['y'])
    dt = np.diff(shot['t'])

    ds = np.sqrt((shot['x'][1:] - shot['x'][:-1])**2 + (shot['y'][1:] - shot['y'][:-1])**2) / 1000
    vel = ds / dt

    ind = np.where(dt > np.mean(dt))[0]

    plt.figure()
    plt.plot(shot['t'], shot['x'], '-')
    plt.plot(shot['t'], shot['y'], '-')
    plt.plot(shot['t'][ind], shot['y'][ind], 'o')
    plt.grid()
    plt.show()

    check = True
    while check:
        # Identify single extremes
        vel_mvmean = np.convolve(vel, np.ones(5)/5, mode='same')  # Moving average
        dev = vel_mvmean - vel
        ind = np.where(dev < -0.5)[0]

        if len(ind) > 0:
            ind = ind[0]
            plt.subplot(3, 1, 1)
            plt.plot(shot['t'][ind], dx[ind], 'o')
            plt.plot(shot['t'][ind], dy[ind], 'o')

            plt.subplot(3, 1, 2)
            plt.plot(shot['t'][ind], vel[ind], 'o')

            plt.subplot(3, 1, 3)
            plt.plot(shot['t'][ind], shot['x'][ind], 'o')
            plt.plot(shot['t'][ind], shot['y'][ind], 'o')

            if abs((dx[ind+1] - dx[ind-1]) / dx[ind-1]) < 0.1 or abs((dy[ind+1] - dy[ind-1]) / dy[ind-1]) < 0.1:
                dx1 = (dx[ind+1] + dx[ind-1]) / 2
                dy1 = (dy[ind+1] + dy[ind-1]) / 2
                shot['x'][ind] = shot['x'][ind-1] + dx1
                shot['y'][ind] = shot['y'][ind-1] + dy1

                dx = np.diff(shot['x'])
                dy = np.diff(shot['y'])

                ds = np.sqrt((shot['x'][1:] - shot['x'][:-1])**2 + (shot['y'][1:] - shot['y'][:-1])**2) / 1000
                vel = ds / dt

                plt.subplot(3, 1, 1)
                plt.plot(shot['t'][ind], dx[ind], 'x')
                plt.plot(shot['t'][ind], dy[ind], 'x')

                plt.subplot(3, 1, 2)
                plt.plot(shot['t'][ind], vel[ind], 'x')

                plt.subplot(3, 1, 3)
                plt.plot(shot['t'][ind], shot['x'][ind], 'x')
                plt.plot(shot['t'][ind], shot['y'][ind], 'x')

                plt.figure()
                plt.plot(shot['t'], shot['y'], '-x')
                plt.show()

    return shot